slash command to update ignored rooms list

Gnuxie opened the pull request:
Based on https://github.com/matrix-org/matrix-js-sdk/pull/2483
Doesn't actually ignore anything, just updates the list in the above PR



<!-- CHANGELOG_PREVIEW_START -->
---
This PR currently has none of the required changelog labels.

A reviewer can add one of: `T-Deprecation`, `T-Enhancement`, `T-Defect`, `T-Task` to indicate what type of change this is, or add `Type: [enhancement/defect/task]` to the description and I'll add them for you.<!-- CHANGELOG_PREVIEW_END -->

Gnuxie committed:
slash command to update ignored rooms list

Based on https://github.com/matrix-org/matrix-js-sdk/pull/2483
Doens't actually ignore anything yet though...

Yoric commented on src/i18n/strings/en_EN.json:
```diff
@@ -428,6 +428,8 @@
     "Stops ignoring a user, showing their messages going forward": "Stops ignoring a user, showing their messages going forward",
     "Unignored user": "Unignored user",
     "You are no longer ignoring %(userId)s": "You are no longer ignoring %(userId)s",
+    "Ignored room": "Ignored room",
```
Note that we're not ignoring the room, only the invites from this room!

Yoric replied:
...actually, you are right, we may be ignoring the room entirely. But to do so, the slash command needs to leave the room immediately, no?

Yoric reviewed (commented):
I think that the syntax should be either
```
/ignore-invite
```

or

```
/ignore invites
```

Otherwise, it's going to be very confusing, since `/ignore` is used for ignoring users entirely.

Also, as you probably know, I cannot approve patches on matrix-react-sdk, so you'll have to find someone from that team to do it once you're ready!

t3chguy commented on src/SlashCommands.tsx:
```diff
@@ -815,6 +815,28 @@ export const Commands = [
                     );
                 }
+                const roomMatches = args.match(/^([!][^:]+:\S+)$/);
+                if (roomMatches) {
```
But you can't write slash commands in a room in which you are holding an invite, you'd need to join the room to the ignore invites and then leave? That doesn't sound safe.

Yoric replied:
Well, user feedback suggests that this is actually a common behavior:

1. Get tricked into a room.
2. Realize that you have been tricked.
3. Decide never to be invited in this room again.
4. ...but we don't have tools for that yet.

We definitely want to add the ability to ignore invites before joining the room, but we're planning to work on that as a later stage, as this will require deeper changes in the UI.

t3chguy replied:
But if you leave before you realise "oh maybe I should ignore future invites" then it'll be too late, you'll have to wait for a second invite, accept it, then finally you can ignore it.

Yoric replied:
Absolutely.

Right now, we're trying to combine two things:
- handle at least one common case with minimal UI;
- lay the foundations for handling the other cases once we have a proper UI design (which we're working on in the background, but with no ETA as of yet).

Gnuxie committed:
Feedback from review

Gnuxie committed:
As requested, change to ignore-invites rather than ignore-room.

The command is much the same but the actual implementation
https://github.com/matrix-org/matrix-js-sdk/pull/2496 will now only
hide invitations.

We now have a command to unignore invitations from a room.

Yoric commented:
@Gnuxie Heads up: the type of `getIgnoredInvites`/`setIgnoredInvites` is changing.

Yoric commented on src/SlashCommands.tsx:
```diff
@@ -850,6 +850,80 @@ export const Commands = [
+            if (Boolean(targetRoomId)) {
+                const ignoredInvites = cli.getIgnoredInvites();
+                if (ignoredInvites.ignored_rooms === undefined) {
+                    ignoredInvites.ignored_rooms = [];
```
In the latest version of the PR, `ignored_rooms` has become `readonly`.

Yoric commented on src/SlashCommands.tsx:
```diff
@@ -850,6 +850,80 @@ export const Commands = [
+                if (!isAlreadyIgnored) {
+                    ignoredInvites.ignored_rooms.push({
+                        room_id: targetRoomId,
+                        ts: Date.now(), // TODO: Check this is the timestamp we want?
```
Yes, it is.

Yoric commented on src/SlashCommands.tsx:
```diff
@@ -850,6 +850,80 @@ export const Commands = [
+                if (roomMatches) {
+                    const roomId = roomMatches[1];
+                    const ignoredInvites = cli.getIgnoredInvites();
+                    if (ignoredInvites.ignored_rooms === undefined) {
```
Same remarks.

andybalaam commented:
@Gnuxie is this abandoned? Will close if so.
